pkg/actuators/machine/actuator.go

```diff
@@ -116,7 +117,37 @@ func (a *Actuator) Create(context context.Context, cluster *clusterv1.Cluster, m
 		}
 		return err
 	}
-	return a.updateStatus(machine, instance)
+	updatedMachine, err := a.updateProviderID(machine, instance)
+	if err != nil {
+		return fmt.Errorf("failed to update machine object with providerID: %v", err)
+	}
+	return a.updateStatus(updatedMachine, instance)
+}
+
+// updateProviderID adds providerID in the machine spec
+func (a *Actuator) updateProviderID(machine *machinev1.Machine, instance *ec2.Instance) (*machinev1.Machine, error) {
+	existingProviderID := machine.Spec.ProviderID
+	machineCopy := machine.DeepCopy()
+	if instance != nil {
+		providerID := fmt.Sprintf("aws:///%s/%s", aws.StringValue(instance.Placement.AvailabilityZone), aws.StringValue(instance.InstanceId))
+
+		if existingProviderID != nil && *existingProviderID == providerID {
+			glog.Infof("ProviderID already set in the machine Spec with value:%s", *existingProviderID)
+			return nil, nil
```

Comment:
Shouldn't this case return the original machine (as passed in)? We go on to use it unconditionally if there is no error:

```go
	updatedMachine, err := a.updateProviderID(machine, instance)
	if err != nil {
		return fmt.Errorf("failed to update machine object with providerID: %v", err)
	}
	return a.updateStatus(updatedMachine, instance)
```